### ISSUE TITLE
Graceful Shutdown

### ISSUE THREAD

**Issue by AndrzejWroGM** [rebus-org/Rebus#1101]
From docs: "With all of the official IoC container integration packages, Rebus will gracefully shut down when the container is disposed". I have issue where after receiving SIGTERM signal my app starts to perform graceful shutdown action and Rebus indeed allows processing already received messages by handlers but, when handlers try send new messages, outgoing pipeline fails because RabbitMqTransport is disposed.
In my case after graceful shutdown I whant my app to finish ongoing conversation and process all already received messages but what's also important and critical in this case I need possibility to send messages. which are direct output from processing "already received" messages by the app. Currently it looks like its not possible.

**Comment by mookid8000** [comment]
That does not sound right. Could you give me some more details about what actually happens? 

I assume you're using the Rebus + Rebus.RabbitMq packages, but which versions?

Which IoC container are you using, and which version of Rebus.*?

Also if you could include the full exception details from the error message it would be great.

**Comment by AndrzejWroGM** [comment]
API in .net 6,
Rebus version 7.1.0
RabbitMq Client package version 6.3.0
Rebus.ServiceProvider

![image](https://github.com/rebus-org/Rebus/assets/24229704/010680ab-dd19-4c74-9630-e46401b914aa)

**Comment by antpata** [comment]
RabbitMQ transport need to call BasicCancel for all consumers as soon as a shutdown is requested and not to dispose channels (models) and connections before message draining is completed.

**Comment by AYss** [comment]
Actually the problem was solved by hooking onto BusLifetimeEvents
```
               configurer.Options(o => o.Decorate(c =>
                {
                    var events = c.Get<BusLifetimeEvents>();

                    events.BusDisposing += () =>
                    {
                        var listener = provider.GetRequiredService<IListener>();
                        listener.StopListening();
                
                        var conversationProcessor = provider.GetService<IConversationProcessor>();
                        conversationProcessor.WaitForSessionsOnShutdown();
                    };

                    return events;
                }));

                return configurer;
```
... where Listener and ConversationProcessor are the tasks that need to do somthing before the bus goes down. I think it is worth adding to the wiki.

**Comment by mookid8000** [comment]
Ok now I've read this a couple of times, and I think I am beginning to understand the issue. But before I do anything, I want to be sure that I'm not way off 😁 

@AYss : Is it correct that the issue you were facing (until you figured out an elegant way to solve it) was that you had some kind of background jobs running that would publish messages via Rebus? And so, you've found out how to make the background jobs stop BEFORE the bus is disposed, so they can finish doing whatever they're doing before the bus is torn down?

@antpata : If I understand you correctly, you are talking about something similar, but at a lower level? I.e. what you're addressing here sounds more like the question of "what should Rebus do with prefetched messagess and messages currently being processed when it has been asked to shut itself down?" I *think* that part has been solved nicely with all transports, because Rebus will by default let handlers run for up to 30 seconds (IIRC) before cancelling them – you can see it in the log as something like "(..) waiting for handler continuations to finish (...)" or something like that. And regarding RabbitMQ, it will automatically bring prefetched messages back to "Ready" when the receiving channel is closed (it sits as "Unacked" when it has been prefetched by the driver).

In summary: @AYss I'm happy that you found a solution, and it would indeed be a nice thing to have on the wiki. I'll add it right away. @antpata To my knowledge Rebus shuts down as it should (letting already executing handlers finish, returning prefetched messages to the "Ready" state). Please let me know if you don't think this is the case.

**Comment by mookid8000** [comment]
@AYss I've added this to the wiki: https://github.com/rebus-org/Rebus/wiki/Startup-shutdown#doing-stuff-before-shutting-down

....and I was also inspired to add the `.Events(...)` thing on the options configurer, making it possible to configure bus lifetime events by going

```csharp
services.AddRebus(
    (configure, provider) => configure
        .(...)
        .Options(o => {
            o.Events(e => {
                // add event listeners in here 🙂
            });
        })
);
```
which is available in Rebus 7.2.0 and 8.0.0-alpha11

**Comment by AYss** [comment]
@mookid8000 Yep, the issue was that the events consumed from devices from outside of our domain (via tcplistener in a background service) were not published into the system before bus went down on shutdown/restart. Blaming handlers `when handlers try send new messages` was a mistake. Thanks for an elegant addition to configurer. Cheers.
